packages/angular_devkit/build_angular/src/utils/routes-extractor/extractor.ts

```diff
@@ -0,0 +1,102 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { ApplicationRef, Injector, Type, createPlatformFactory, platformCore } from '@angular/core';
+import {
+  INITIAL_CONFIG,
+  ɵINTERNAL_SERVER_PLATFORM_PROVIDERS as INTERNAL_SERVER_PLATFORM_PROVIDERS,
+} from '@angular/platform-server';
+import { Route, Router, ɵROUTER_PROVIDERS } from '@angular/router';
+import { first } from 'rxjs/operators'; // Import from `/operators` to support rxjs 6 which is still supported by the Framework.
+
+// TODO(alanagius): replace the below once `RouterConfigLoader` is privately exported from `@angular/router`.
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const RouterConfigLoader = ɵROUTER_PROVIDERS[5] as any;
+type RouterConfigLoader = typeof RouterConfigLoader;
+
+async function* getRoutesFromRouterConfig(
+  routes: Route[],
+  routerConfigLoader: RouterConfigLoader,
+  injector: Injector,
+  parent = '',
+): AsyncIterableIterator<string> {
+  for (const route of routes) {
+    const { path, redirectTo, loadChildren, children } = route;
+    if (path === undefined) {
+      continue;
+    }
+
+    if (redirectTo !== undefined) {
+      // TODO: handle `redirectTo`.
+      continue;
+    }
+
+    if (/[:*]/.test(path)) {
+      // TODO: handle parameterized routes population.
+      continue;
+    }
+
+    const currentRoutePath = buildRoutePath(parent, path);
+    yield currentRoutePath;
+
+    if (children?.length || loadChildren) {
+      yield* getRoutesFromRouterConfig(
+        children ?? (await routerConfigLoader.loadChildren(injector, route).toPromise()).routes,
+        routerConfigLoader,
+        injector,
+        currentRoutePath,
+      );
+    }
+  }
+}
+
+export async function* extractRoutes(
+  bootstrapAppFnOrModule: (() => Promise<ApplicationRef>) | Type<unknown>,
+  document: string,
+): AsyncIterableIterator<string> {
+  const platformRef = createPlatformFactory(platformCore, 'server', [
+    [
+      {
+        provide: INITIAL_CONFIG,
+        useValue: { document, url: '' },
+      },
+    ],
+    ...INTERNAL_SERVER_PLATFORM_PROVIDERS,
+  ])();
+
+  try {
+    let applicationRef: ApplicationRef;
+    if (isBootstrapFn(bootstrapAppFnOrModule)) {
+      applicationRef = await bootstrapAppFnOrModule();
+    } else {
+      const moduleRef = await platformRef.bootstrapModule(bootstrapAppFnOrModule);
+      applicationRef = moduleRef.injector.get(ApplicationRef);
+    }
+
+    // Wait until the application is stable.
+    await applicationRef.isStable.pipe(first((isStable) => isStable)).toPromise();
```

Comment:
In the future, it would be nice if we had a promise-based API alternative here. Something like `untilStable()` or similar. No piping and also no operator imports then.

Reply:
Yeah, we definitely